{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a04cd0f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import matplotlib.patches as mpatches\n",
    "from pywaffle import Waffle\n",
    "\n",
    "# Read the CSV file into a DataFrame\n",
    "data = pd.read_csv('../../Data/Fig3/Talos_solves_VCGS-prospective_261128.csv', header=0, index_col=[1, 0], usecols=range(1, 7))\n",
    "\n",
    "# Combine duplicate rows by summing their values\n",
    "# This is needed because the data breaks out some additional categories that we are not interested in\n",
    "data_combined = data.groupby(level=[0, 1]).sum()\n",
    "\n",
    "\n",
    "# colors_list = ['#EC4E20',\n",
    "#         '#F2805F',\n",
    "#         '#FCCCBE',\n",
    "#         '#016FB9',\n",
    "#         '#6BAFDC',\n",
    "#         '#218380',\n",
    "#         '#57ABA9',\n",
    "#         '#8CD4D1',\n",
    "#         '#C2FCFA',\n",
    "#         '#51514B',\n",
    "#         '#A3A39A']\n",
    "\n",
    "colors_list = ['#EC4E20', #flame orginal\n",
    "        '#F69A7F', #flame light\n",
    "        '#FCCCBE', #flame lightest\n",
    "        '#FF9505', #orange original\n",
    "        '#FFC474', #orange light\n",
    "        '#016FB9', #azul original\n",
    "        '#4799D0', #azul light\n",
    "        '#8EC4E8', #azul lighter\n",
    "        '#D4EEFF', #azul lightest\n",
    "        '#57ABA9', #teal light\n",
    "        '#C2FCFA'] #teal lightest\n",
    "\n",
    "#print([f\"{k} ({v})\" for k, v in data['New association'].items()][0:3])\n",
    "\n",
    "fig = plt.figure(\n",
    "    FigureClass=Waffle,\n",
    "    plots={\n",
    "        411: {\n",
    "            'values': data_combined.loc[('Gene-disease Associations'),'All'],\n",
    "            'colors': colors_list[0:3],\n",
    "            #'title': {'label': 'Gene-disease associations', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        412: {\n",
    "            'values': data_combined.loc[('Variant Reclassifications'),'All'],\n",
    "            'colors': colors_list[3:5],\n",
    "            #'title': {'label': 'Variant reclassifications', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        413: {\n",
    "            'values': data_combined.loc[('Missed in Manual Analysis'),'All'],\n",
    "            'colors': colors_list[5:9],\n",
    "            #'title': {'label': 'Missed in original analysis', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        414: {\n",
    "            'values': data_combined.loc[('CNV/SV'),'All'],\n",
    "            'colors': colors_list[9:12],\n",
    "            #'title': {'label': 'CNV/SV', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "    },\n",
    "    rows=5,\n",
    "    #colors=colors_list,\n",
    "    rounding_rule='ceil',\n",
    "    figsize=(5, 5)\n",
    ")\n",
    "\n",
    "# import matplotlib.patches as mpatches\n",
    "# import matplotlib.pyplot as plt\n",
    "\n",
    "# patches = []\n",
    "# for x in range(len(solve_data)):\n",
    "#     patches.append(mpatches.Patch(color=colors_list[x], label=list(solve_data.keys())[x]))\n",
    "# plt.legend(handles=patches, bbox_to_anchor=(3,2), loc=\"lower left\")\n",
    "\n",
    "# legend1 = plt.legend(handles=patches[0:3], bbox_to_anchor=(1.5,5.7), loc=\"upper left\")\n",
    "# legend2 = plt.legend(handles=patches[3:5], bbox_to_anchor=(1.5,4.3), loc=\"upper left\")\n",
    "# legend3 = plt.legend(handles=patches[5:9], bbox_to_anchor=(1.5,2.8), loc=\"upper left\")\n",
    "# legend4 = plt.legend(handles=patches[9:11], bbox_to_anchor=(1.5,1.1), loc=\"upper left\")\n",
    "# fig.add_artist(legend1)\n",
    "# fig.add_artist(legend2)\n",
    "# fig.add_artist(legend3)\n",
    "# fig.add_artist(legend4)\n",
    "\n",
    "\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2.png')\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea670a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import matplotlib.patches as mpatches\n",
    "from pywaffle import Waffle\n",
    "\n",
    "# Read the CSV file into a DataFrame\n",
    "data = pd.read_csv('../../Data/Fig3/Talos_solves_VCGS-prospective_261128.csv', header=0, index_col=[1, 0], usecols=range(1, 7))\n",
    "\n",
    "# Combine duplicate rows by summing their values\n",
    "# This is needed because the data breaks out some additional categories that we are not interested in\n",
    "data_combined = data.groupby(level=[0, 1]).sum()\n",
    "\n",
    "\n",
    "# colors_list = ['#EC4E20',\n",
    "#         '#F2805F',\n",
    "#         '#FCCCBE',\n",
    "#         '#016FB9',\n",
    "#         '#6BAFDC',\n",
    "#         '#218380',\n",
    "#         '#57ABA9',\n",
    "#         '#8CD4D1',\n",
    "#         '#C2FCFA',\n",
    "#         '#51514B',\n",
    "#         '#A3A39A']\n",
    "\n",
    "colors_list = ['#EC4E20', #flame orginal\n",
    "        '#F69A7F', #flame light\n",
    "        '#FCCCBE', #flame lightest\n",
    "        '#FF9505', #orange original\n",
    "        '#FFC474', #orange light\n",
    "        '#016FB9', #azul original\n",
    "        '#4799D0', #azul light\n",
    "        '#8EC4E8', #azul lighter\n",
    "        '#D4EEFF', #azul lightest\n",
    "        '#57ABA9', #teal light\n",
    "        '#C2FCFA'] #teal lightest\n",
    "\n",
    "#print([f\"{k} ({v})\" for k, v in data['New association'].items()][0:3])\n",
    "\n",
    "fig = plt.figure(\n",
    "    FigureClass=Waffle,\n",
    "    plots={\n",
    "        411: {\n",
    "            'values': data_combined.loc[('Gene-disease Associations'),'NDD'],\n",
    "            'colors': colors_list[0:3],\n",
    "            #'title': {'label': 'Gene-disease associations', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        412: {\n",
    "            'values': data_combined.loc[('Variant Reclassifications'),'NDD'],\n",
    "            'colors': colors_list[3:5],\n",
    "            #'title': {'label': 'Variant reclassifications', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        413: {\n",
    "            'values': data_combined.loc[('Missed in Manual Analysis'),'NDD'],\n",
    "            'colors': colors_list[5:9],\n",
    "            #'title': {'label': 'Missed in original analysis', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        414: {\n",
    "            'values': data_combined.loc[('CNV/SV'),'NDD'],\n",
    "            'colors': colors_list[9:12],\n",
    "            #'title': {'label': 'CNV/SV', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "    },\n",
    "    rows=5,\n",
    "    #colors=colors_list,\n",
    "    rounding_rule='ceil',\n",
    "    figsize=(5, 5)\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2_NDD.png')\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2_NDD.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19b966ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import matplotlib.patches as mpatches\n",
    "from pywaffle import Waffle\n",
    "\n",
    "# Read the CSV file into a DataFrame\n",
    "data = pd.read_csv('../../Data/Fig3/Talos_solves_VCGS-prospective_261128.csv', header=0, index_col=[1, 0], usecols=range(1, 7))\n",
    "\n",
    "# Combine duplicate rows by summing their values\n",
    "# This is needed because the data breaks out some additional categories that we are not interested in\n",
    "data_combined = data.groupby(level=[0, 1]).sum()\n",
    "\n",
    "\n",
    "# colors_list = ['#EC4E20',\n",
    "#         '#F2805F',\n",
    "#         '#FCCCBE',\n",
    "#         '#016FB9',\n",
    "#         '#6BAFDC',\n",
    "#         '#218380',\n",
    "#         '#57ABA9',\n",
    "#         '#8CD4D1',\n",
    "#         '#C2FCFA',\n",
    "#         '#51514B',\n",
    "#         '#A3A39A']\n",
    "\n",
    "colors_list = ['#EC4E20', #flame orginal\n",
    "        '#F69A7F', #flame light\n",
    "        '#FCCCBE', #flame lightest\n",
    "        '#FF9505', #orange original\n",
    "        '#FFC474', #orange light\n",
    "        '#016FB9', #azul original\n",
    "        '#4799D0', #azul light\n",
    "        '#8EC4E8', #azul lighter\n",
    "        '#D4EEFF', #azul lightest\n",
    "        '#57ABA9', #teal light\n",
    "        '#C2FCFA'] #teal lightest\n",
    "\n",
    "#print([f\"{k} ({v})\" for k, v in data['New association'].items()][0:3])\n",
    "\n",
    "fig = plt.figure(\n",
    "    FigureClass=Waffle,\n",
    "    plots={\n",
    "        411: {\n",
    "            'values': data_combined.loc[('Gene-disease Associations'),'Cardiac'],\n",
    "            'colors': colors_list[0:3],\n",
    "            #'title': {'label': 'Gene-disease associations', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        412: {\n",
    "            'values': data_combined.loc[('Variant Reclassifications'),'Cardiac'],\n",
    "            'colors': colors_list[3:5],\n",
    "            #'title': {'label': 'Variant reclassifications', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        413: {\n",
    "            'values': data_combined.loc[('Missed in Manual Analysis'),'Cardiac'],\n",
    "            'colors': colors_list[5:9],\n",
    "            #'title': {'label': 'Missed in original analysis', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        414: {\n",
    "            'values': data_combined.loc[('CNV/SV'),'Cardiac'],\n",
    "            'colors': colors_list[9:12],\n",
    "            #'title': {'label': 'CNV/SV', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "    },\n",
    "    rows=5,\n",
    "    #colors=colors_list,\n",
    "    rounding_rule='ceil',\n",
    "    figsize=(5, 5)\n",
    ")\n",
    "\n",
    "\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2_Cardiac.png')\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2_Cardiac.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f644c94",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import matplotlib.patches as mpatches\n",
    "from pywaffle import Waffle\n",
    "\n",
    "# Read the CSV file into a DataFrame\n",
    "data = pd.read_csv('../../Data/Fig3/Talos_solves_VCGS-prospective_261128.csv', header=0, index_col=[1, 0], usecols=range(1, 7))\n",
    "\n",
    "# Combine duplicate rows by summing their values\n",
    "# This is needed because the data breaks out some additional categories that we are not interested in\n",
    "data_combined = data.groupby(level=[0, 1]).sum()\n",
    "\n",
    "# Sorting dataframe so that legends appear in the desired order\n",
    "custom_sort={'Gene-disease Associations': 1, 'Variant Reclassifications': 2, 'Missed in Manual Analysis': 3, 'CNV/SV': 4}\n",
    "data_sorted = data_combined.sort_index(key=lambda x: x.map(custom_sort))\n",
    "\n",
    "colors_list = ['#EC4E20', #flame orginal\n",
    "        '#F69A7F', #flame light\n",
    "        '#FCCCBE', #flame lightest\n",
    "        '#FF9505', #orange original\n",
    "        '#FFC474', #orange light\n",
    "        '#016FB9', #azul original\n",
    "        '#4799D0', #azul light\n",
    "        '#8EC4E8', #azul lighter\n",
    "        '#D4EEFF', #azul lightest\n",
    "        '#57ABA9', #teal light\n",
    "        '#C2FCFA'] #teal lightest\n",
    "\n",
    "#print([f\"{k} ({v})\" for k, v in data['New association'].items()][0:3])\n",
    "\n",
    "fig = plt.figure(\n",
    "    FigureClass=Waffle,\n",
    "    plots={\n",
    "        411: {\n",
    "            'values': data_combined.loc[('Gene-disease Associations'),'Renal'],\n",
    "            'colors': colors_list[0:3],\n",
    "            #'title': {'label': 'Gene-disease associations', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        412: {\n",
    "            'values': data_combined.loc[('Variant Reclassifications'),'Renal'],\n",
    "            'colors': colors_list[3:5],\n",
    "            #'title': {'label': 'Variant reclassifications', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        413: {\n",
    "            'values': data_combined.loc[('Missed in Manual Analysis'),'Renal'],\n",
    "            'colors': colors_list[5:9],\n",
    "            #'title': {'label': 'Missed in original analysis', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "        414: {\n",
    "            'values': data_combined.loc[('CNV/SV'),'Renal'],\n",
    "            'colors': colors_list[9:12],\n",
    "            #'title': {'label': 'CNV/SV', 'loc': 'left', 'fontsize': 12}\n",
    "        },\n",
    "    },\n",
    "    rows=5,\n",
    "    #colors=colors_list,\n",
    "    rounding_rule='ceil',\n",
    "    figsize=(5, 5)\n",
    ")\n",
    "\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "patches = []\n",
    "for x in range(len(data_sorted.index.get_level_values(1))):\n",
    "    labels = list(data_sorted.index.get_level_values(1))\n",
    "    patches.append(mpatches.Patch(color=colors_list[x], label=labels[x]))\n",
    "plt.legend(handles=patches, bbox_to_anchor=(3,2), loc=\"lower left\")\n",
    "\n",
    "legend1 = plt.legend(handles=patches[0:3], bbox_to_anchor=(1.5,4.3), loc=\"upper left\")\n",
    "legend2 = plt.legend(handles=patches[3:5], bbox_to_anchor=(1.5,3.0), loc=\"upper left\")\n",
    "legend3 = plt.legend(handles=patches[5:9], bbox_to_anchor=(1.5,2.1), loc=\"upper left\")\n",
    "legend4 = plt.legend(handles=patches[9:11], bbox_to_anchor=(1.5,0.8), loc=\"upper left\")\n",
    "fig.add_artist(legend1)\n",
    "fig.add_artist(legend2)\n",
    "fig.add_artist(legend3)\n",
    "fig.add_artist(legend4)\n",
    "\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2_Renal.png')\n",
    "fig.savefig('../../Figures/Fig3/Fig3_panel2_Renal.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "353895bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## This version rasterises all inputs so will try a different approach to combine the panels\n",
    "\n",
    "# import matplotlib.pyplot as plt\n",
    "# import matplotlib.image as mpimg\n",
    "\n",
    "# # Paths to the images\n",
    "# image_paths = ['../../Figures/Fig3/Fig3_panel2.png', '../../Figures/Fig3/Fig3_panel2_NDD.png', '../../Figures/Fig3/Fig3_panel2_Cardiac.png', '../../Figures/Fig3/Fig3_panel2_Renal.png']\n",
    "\n",
    "# # Set the default font to Arial\n",
    "# plt.rcParams['font.family'] = 'Arial'\n",
    "\n",
    "# # Create a figure with three subplots arranged horizontally with variable widths\n",
    "# fig = plt.figure(figsize=(18, 6))\n",
    "\n",
    "# # Manually add subplots with specific positions\n",
    "# positions = [\n",
    "#     [0.25, 0.1, 0.25, 0.75],  # [left, bottom, width, height] for the first subplot\n",
    "#     [0.47, 0.1, 0.25, 0.75],  # Adjust the left position for the second subplot\n",
    "#     [0.58, 0.1, 0.25, 0.75],  # Adjust the left position for the third subplot\n",
    "#     [0.69, 0.1, 0.25, 0.75]  # Adjust the left position for the third subplot\n",
    "# ]\n",
    "\n",
    "# # Load and display the images in each subplot\n",
    "# for i, pos in enumerate(positions):\n",
    "#     ax = fig.add_axes(pos)\n",
    "#     img = mpimg.imread(image_paths[i])\n",
    "#     ax.imshow(img, aspect='equal')\n",
    "#     ax.axis('off')  # Turn off the axis\n",
    "\n",
    "# # Adjust layout to prevent overlap\n",
    "# #plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "\n",
    "# # Add text that spans across subplot boundaries\n",
    "# fig.text(0.205, 0.86, 'New Diagnoses\\n(Count)', ha='center', fontsize=12, fontweight='bold')\n",
    "# fig.text(0.31, 0.86, 'Full Cohort\\n(248)', ha='center', fontsize=12)\n",
    "# fig.text(0.510, 0.85, 'NDD\\nSub-cohort\\n(91)', ha='center', fontsize=12)\n",
    "# fig.text(0.620, 0.85, 'Cardiac\\nSub-cohort\\n(54)', ha='center', fontsize=12)\n",
    "# fig.text(0.720, 0.85, 'Renal\\nSub-cohort\\n(43)', ha='center', fontsize=12)\n",
    "\n",
    "# fig.text(0.205, 0.73, 'Gene-disease\\nAssociations', ha='center', fontsize=12)\n",
    "# fig.text(0.205, 0.55, 'Variant\\nReclassifications', ha='center', fontsize=12)\n",
    "# fig.text(0.205, 0.37, 'Missed in\\nManual Analysis', ha='center', fontsize=12)\n",
    "# fig.text(0.205, 0.195, 'CNV/SV', ha='center', fontsize=12)\n",
    "\n",
    "# fig.text(0.205, 0.04, 'Total Cohort Size\\nNew Diagnosis %', ha='center', fontsize=12, fontweight='bold')\n",
    "# fig.text(0.31, 0.04, 'N=4738\\n5.2%', ha='center', fontsize=12)\n",
    "# fig.text(0.510, 0.04, 'N=1665\\n5.4%', ha='center', fontsize=12)\n",
    "# fig.text(0.620, 0.04, 'N=997\\n5.4%', ha='center', fontsize=12)\n",
    "# fig.text(0.720, 0.04, 'N=695\\n6.1%', ha='center', fontsize=12)\n",
    "\n",
    "# fig.savefig('../../Figures/Fig3/Fig3_panel2_All.png')\n",
    "# fig.savefig('../../Figures/Fig3/Fig3_panel2_All.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "025fcf22",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pypdf import PdfReader, PdfWriter\n",
    "\n",
    "panel_paths = [\n",
    "    \"../../Figures/Fig3/Fig3_panel2.pdf\",\n",
    "    \"../../Figures/Fig3/Fig3_panel2_NDD.pdf\",\n",
    "    \"../../Figures/Fig3/Fig3_panel2_Cardiac.pdf\",\n",
    "    \"../../Figures/Fig3/Fig3_panel2_Renal.pdf\",\n",
    "]\n",
    "pages = [PdfReader(p).pages[0] for p in panel_paths]\n",
    "\n",
    "# Calculate original size and add margin (e.g., 15% top/bottom, 10% left/right)\n",
    "orig_w = sum(p.mediabox.width for p in pages)\n",
    "orig_h = max(p.mediabox.height for p in pages)\n",
    "margin_x = orig_w * 0.10\n",
    "margin_y = orig_h * 0.15\n",
    "new_w = orig_w + 2 * margin_x\n",
    "new_h = orig_h + 2 * margin_y\n",
    "\n",
    "writer = PdfWriter()\n",
    "page = writer.add_blank_page(width=new_w, height=new_h)\n",
    "\n",
    "x = margin_x\n",
    "y = margin_y  # place panels above bottom margin; adjust if you want more top space\n",
    "for p in pages:\n",
    "    page.merge_translated_page(p, tx=x, ty=y, expand=False)\n",
    "    x += p.mediabox.width\n",
    "\n",
    "combo_path = \"../../Figures/Fig3/Fig3_panel2_All.pdf\"\n",
    "with open(combo_path, \"wb\") as f:\n",
    "    writer.write(f)\n",
    "\n",
    "\n",
    "from reportlab.pdfgen import canvas\n",
    "import tempfile, os\n",
    "from pypdf import PdfReader, PdfWriter\n",
    "\n",
    "overlay_file = tempfile.NamedTemporaryFile(delete=False, suffix=\".pdf\")\n",
    "c = canvas.Canvas(overlay_file.name, pagesize=(new_w, new_h))\n",
    "c.setFont(\"Helvetica-Bold\", 12)\n",
    "\n",
    "def rel_text(x_rel, y_rel, text, bold=False, size=12):\n",
    "    c.setFont(\"Helvetica-Bold\" if bold else \"Helvetica\", size)\n",
    "    c.drawCentredString(x_rel * new_w, y_rel * new_h, text)\n",
    "\n",
    "def rel_multiline(x_rel, y_rel, text, line_height=14, bold=False, size=12):\n",
    "    c.setFont(\"Helvetica-Bold\" if bold else \"Helvetica\", size)\n",
    "    lines = text.splitlines()\n",
    "    # draw top-aligned; adjust offset if you want center/other alignment\n",
    "    for i, line in enumerate(lines):\n",
    "        c.drawCentredString(x_rel * new_w, (y_rel - i * line_height / new_h) * new_h, line)\n",
    "\n",
    "# Use relative coords that land in the added whitespace (example positions)\n",
    "rel_multiline(0.05, 0.92, \"New Diagnoses\\n(Count)\", bold=True)\n",
    "rel_multiline(0.17,  0.92, \"Full Cohort\\n(248)\")\n",
    "rel_multiline(0.33, 0.92, \"NDD Sub-cohort\\n(91)\")\n",
    "rel_multiline(0.53, 0.92, \"Cardiac Sub-cohort\\n(54)\")\n",
    "rel_multiline(0.725, 0.92, \"Renal Sub-cohort\\n(43)\")\n",
    "rel_multiline(0.05, 0.79, \"Gene-disease\\nAssociations\")\n",
    "rel_multiline(0.05, 0.60, \"Variant\\nReclassifications\")\n",
    "rel_multiline(0.05, 0.41, \"Missed in\\nManual Analysis\")\n",
    "rel_multiline(0.05, 0.21, \"CNV/SV\")\n",
    "rel_multiline(0.05, 0.09, \"Total Cohort Size\\nNew Diagnosis %\", bold=True)\n",
    "rel_multiline(0.17,  0.09, \"N=4738\\n5.2%\")\n",
    "rel_multiline(0.33, 0.09, \"N=1665\\n5.4%\")\n",
    "rel_multiline(0.53, 0.09, \"N=997\\n5.4%\")\n",
    "rel_multiline(0.725, 0.09, \"N=695\\n6.1%\")\n",
    "\n",
    "c.save()\n",
    "\n",
    "# Merge overlay onto the combined page\n",
    "base = PdfReader(combo_path)\n",
    "overlay = PdfReader(overlay_file.name)\n",
    "out = PdfWriter()\n",
    "pg = base.pages[0]\n",
    "pg.merge_page(overlay.pages[0])\n",
    "out.add_page(pg)\n",
    "with open(combo_path, \"wb\") as f:\n",
    "    out.write(f)\n",
    "os.unlink(overlay_file.name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "869ea925",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d05283cb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Talos figs",
   "language": "python",
   "name": "talos_figs"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
